core/src/autogluon/core/models/abstract/abstract_model.py

```diff
@@ -573,7 +573,15 @@ def _preprocess_fit_resources(self, silent=False, total_resources=None, parallel
                 )
                 minimum_model_num_cpus = minimum_model_resources.get('num_cpus', 1)
                 enforced_num_cpus = max(minimum_model_num_cpus, enforced_num_cpus - 2)  # leave some cpu resources for process running by cluster nodes
-                kwargs["num_cpus"] = enforced_num_cpus
+            max_resources = self._get_maximum_resources()
+            max_num_cpus = max_resources.get("num_cpus", None)
+            max_num_gpus = max_resources.get("num_gpus", None)
+            if max_num_cpus is not None:
+                enforced_num_cpus = min(max_num_cpus, enforced_num_cpus)
+            if max_num_gpus is not None:
+                enforced_num_gpus = min(max_num_gpus, enforced_num_gpus)
+            kwargs["num_cpus"] = enforced_num_cpus
+            kwargs["num_gpus"] = enforced_num_gpus
```

Comment:
the logic starting at line 578 should be moved to its own method, currently I'm pretty confused trying to understand in what situations line 578+ is called and in which situations the earlier code is called.

Reply:
Logic starting 578+ always called no matter distributed or not

Reply:
Maybe line numbers are not best way of saying it, I mean the code after `return kwargs`

Starting with

```python3
resource_manager = get_resource_manager()
system_num_cpus = resource_manager.get_cpu_count()
...
```